pkg/models/kubeconfig/kubeconfig.go

```diff
@@ -21,295 +21,228 @@ package kubeconfig
-		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
-		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
-		EmailAddresses:        info.EmailAddress,
-	}
-}
 
-func generateCaAndKey(user, caPath, keyPath string) (string, string, error) {
-	crtInfo := CertInformation{CommonName: user, IsCA: false}
+	// release csr, if it fails need to delete it manually
+	defer func() {
+		k8sClient.CertificatesV1beta1().CertificateSigningRequests().Delete(csrName, &metav1.DeleteOptions{})
+	}()
 
-	crt, pri, err := Parse(caPath, keyPath)
-	if err != nil {
-		klog.Error(err)
-		return "", "", err
-	}
-	cert, key, err := createCRT(crt, pri, crtInfo)
-	if err != nil {
-		klog.Error(err)
-		return "", "", err
+	k8sCSR.Status = certificatesv1beta1.CertificateSigningRequestStatus{
+		Conditions: []certificatesv1beta1.CertificateSigningRequestCondition{{
+			Type:    "Approved",
+			Reason:  "KubeSphereApprove",
+			Message: "This CSR was approved by KubeSphere certificate approve.",
+			LastUpdateTime: metav1.Time{
+				Time: time.Now(),
+			},
+		}},
 	}
 
-	base64Cert := base64.StdEncoding.EncodeToString(cert)
-	base64Key := base64.StdEncoding.EncodeToString(key)
-	return base64Cert, base64Key, nil
-}
+	// approve csr
+	k8sCSR, err = k8sClient.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(k8sCSR)
 
-func createKubeConfig(username string) (string, error) {
-	tmpKubeConfig := kubeConfig{ApiVersion: "v1", Kind: "Config"}
-	serverCa, err := ioutil.ReadFile(caPath)
 	if err != nil {
 		klog.Errorln(err)
-		return "", err
+		return nil, err
 	}
-	base64ServerCa := base64.StdEncoding.EncodeToString(serverCa)
-	tmpClusterInfo := clusterInfo{CertificateAuthorityData: base64ServerCa, Server: client.ClientSets().K8s().Master()}
-	tmpCluster := cluster{Cluster: tmpClusterInfo, Name: clusterName}
-	tmpKubeConfig.Clusters = append(tmpKubeConfig.Clusters, tmpCluster)
-
-	contextName := username + "@" + clusterName
-	tmpContext := contextObject{Context: contextInfo{User: username, Cluster: clusterName, NameSpace: defaultNamespace}, Name: contextName}
-	tmpKubeConfig.Contexts = append(tmpKubeConfig.Contexts, tmpContext)
 
-	cert, key, err := generateCaAndKey(username, caPath, keyPath)
+	// get client cert
+	var cert []byte
+	maxRetries := 3
+	for i := 0; i < maxRetries; i++ {
 
-	if err != nil {
-		return "", err
-	}
+		k8sCSR, err = k8sClient.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
 
-	tmpUser := user{User: userInfo{CaData: cert, KeyData: key}, Name: username}
-	tmpKubeConfig.Users = append(tmpKubeConfig.Users, tmpUser)
-	tmpKubeConfig.CurrentContext = contextName
+		if k8sCSR != nil && k8sCSR.Status.Certificate != nil {
+			cert = k8sCSR.Status.Certificate
+			break
+		}
 
-	config, err := yaml.Marshal(tmpKubeConfig)
-	if err != nil {
-		return "", err
+		// sleep 0/1/2 second
+		time.Sleep(time.Second * time.Duration(i))
```

Comment:
this is not a good practice, the api here is synchronous, it means it will take at most 3 seconds for this api to be completed and it may still failed.  I think either 1) build this logic into a controller 2) lower the sleep duration. I prefer the first solution, the second is temporary .

Reply:
This part of the code will be moved to the controller in 3.0. It's just a workaround for 2.1.1, kubeconfig can be recreate even if it fails.